src/resources/class-modifiers-for-apis.md

```diff
@@ -0,0 +1,359 @@
+
+[dangerous and fragile]: https://en.wikipedia.org/wiki/Fragile_base_class
+
+Marking the class `interface` lets users construct it (unless it's also marked
+`abstract`) and implement the class's interface,
+but prevents them from reusing any of its code.
+
+When a class is marked `interface`, the restriction can be ignored within
+the library where the class is declared.
+Inside the library, you're free to extend it since it's all your code
+and presumably you know what you're doing.
+The restriction applies to other packages,
+and even other libraries within your own package.
+
+## The `base` modifier
+
+The `base` modifier is sort of the opposite of `interface`:
+it allows you to use the class in an `extends` clause
+or use a mixin or mixin class in a `with` clause,
+but disallows code outside of the class's library
+from using the class or mixin in an `implements` clause.
+
+This ensures that every object that is an instance
+of your class or mixin's interface inherits your actual implementation.
+In particular, this means that every instance will include
+all of the private members your class or mixin declares.
+This can help prevent runtime errors that might otherwise occur.
+
+Consider this library:
+
+```dart
+// a.dart
+class A {
+  void _privateMethod() {
+    print('I inherited from A');
+  }
+}
+
+void callPrivateMethod(A a) {
+  a._privateMethod();
+}
+```
+
+This code seems fine on its own,
+but there's nothing preventing a user from creating another library like this:
+
+```dart
+// b.dart
+import 'a.dart';
+
+class B implements A {
+  // No implementation of _privateMethod()!
+}
+
+main() {
+  callPrivateMethod(B()); // Runtime exception!
+}
+```
+
+Adding the `base` modifier to the class can help prevent these runtime errors.
+As with `interface`, you can ignore this restriction
+in the same library where the `base` class or mixin is declared,
+and then subclasses in the same library
+will be reminded to implement the private methods.
+But note that the next section *does* apply:
+
+### Base transitivity
+
+The goal of marking a class `base` is to ensure that
+every instance of that type concretely inherits from it.
+To maintain this, the base restriction is "contagious".
+Every subtype of a type marked `base` -- *direct or indirect* --
+must also prevent being implemented too.
+That means it must be marked `base`
+(or `final` or `sealed`, which we'll get to next).
+
+Applying `base` to a type requires some care then,
+because it affects not just what users can do with your class or mixin,
+but also the affordances *their* subclasses can offer.
+Once you've put `base` on a type, the whole hierarchy under it
+is prohibited from being implemented.
+
+That sounds intense, but it's how most other programming languages
+have always worked.
+Most don't have implicit interfaces at all,
+so when you declare a class in Java, C#, or other languages,
+you effectively have the same limitation.
```

Comment:
Perhaps call it a 'constraint' rather than a 'limitation': It's actually fair to say that it is a limitation in those other languages, but in the context of Dart it is a deliberate choice: "I want the complete package consisting of some guarantees and some constraints", and if the constraint/limitation part is a problem then just don't write `base`.